src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/FrameworkParametersCompletionProvider.cs

```diff
@@ -0,0 +1,594 @@
+    {
+        var current = node;
+        while (current != null)
+        {
+            if (current.IsKind(SyntaxKind.Parameter))
+            {
+                return current;
+            }
+
+            current = current.Parent;
+        }
+
+        return null;
+    }
+
+    private static SyntaxNode? TryFindMinimalApiArgument(SyntaxNode node)
+    {
+        var current = node;
+        while (current != null)
+        {
+            if (current.Parent?.IsKind(SyntaxKind.Argument) ?? false)
+            {
+                if (current.Parent?.Parent?.IsKind(SyntaxKind.ArgumentList) ?? false)
+                {
+                    return current;
+                }
+            }
+
+            current = current.Parent;
+        }
+
+        return null;
+    }
+
+    private static bool HasNonRouteAttribute(SyntaxToken token, SemanticModel semanticModel, WellKnownTypes wellKnownTypes, CancellationToken cancellationToken)
+    {
+        if (token.Parent?.Parent is ParameterSyntax parameter)
+        {
+            foreach (var attributeList in parameter.AttributeLists.OfType<AttributeListSyntax>())
+            {
+                foreach (var attribute in attributeList.Attributes)
+                {
+                    var attributeTypeSymbol = semanticModel.GetSymbolInfo(attribute, cancellationToken).GetAnySymbol();
+
+                    if (attributeTypeSymbol != null)
+                    {
+                        foreach (var nonRouteMetadataType in wellKnownTypes.NonRouteMetadataTypes)
+                        {
+                            if (attributeTypeSymbol.ContainingSymbol is ITypeSymbol typeSymbol &&
+                                typeSymbol.Implements(nonRouteMetadataType))
+                            {
+                                return true;
+                            }
+                        }
+                        if (SymbolEqualityComparer.Default.Equals(attributeTypeSymbol.ContainingSymbol, wellKnownTypes.AsParametersAttribute))
+                        {
+                            return true;
+                        }
+                    }
+                }
+            }
+        }
+
+        return false;
+    }
+
+    private static bool IsCurrentParameterBindable(SyntaxToken token, SemanticModel semanticModel, CancellationToken cancellationToken)
+    {
+        if (token.Parent.IsKind(SyntaxKind.PredefinedType))
+        {
+            return true;
+        }
+
+        var parameterTypeSymbol = semanticModel.GetSymbolInfo(token.Parent, cancellationToken).GetAnySymbol();
+        if (parameterTypeSymbol is INamedTypeSymbol typeSymbol)
+        {
+            if (typeSymbol.SpecialType == SpecialType.System_String)
+            {
+                return true;
+            }
+
+            // Check if the parameter type has a static TryParse method.
+            foreach (var item in typeSymbol.GetMembers("TryParse"))
```

Comment:
Would it be possible to use https://github.com/dotnet/aspnetcore/blob/main/src/Shared/ParameterBindingMethodCache.cs#L57 so that we won't need to remember to update this when we make improvements, like supporting `IParsable<TSelf>`. Additionally, `bool TryParse(string, IFormatProvider, out T)` is supported, but not handled here.

Reply:
> Would it be possible to use [`main`/src/Shared/ParameterBindingMethodCache.cs#L57](https://github.com/dotnet/aspnetcore/blob/main/src/Shared/ParameterBindingMethodCache.cs?rgh-link-date=2022-10-18T17%3A16%3A08Z#L57) so that we won't need to remember to update this when we make improvements, like supporting `IParsable<TSelf>`.

No, this isn't reflection. Roslyn uses `ITypeSymbol`, not `Type`.

> Additionally, `bool TryParse(string, IFormatProvider, out T)` is supported, but not handled here.

Ok!